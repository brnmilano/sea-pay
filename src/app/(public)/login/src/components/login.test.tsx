import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Login from "./login";

// Mock do react-hot-toast
jest.mock("react-hot-toast", () => ({
  __esModule: true,
  default: {
    success: jest.fn(),
    error: jest.fn(),
  },
}));

// Mock do next navigation
jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      push: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn(),
      back: jest.fn(),
      pathname: "/",
      query: {},
      asPath: "/",
    };
  },
  usePathname() {
    return "/";
  },
  useSearchParams() {
    return new URLSearchParams();
  },
}));

describe("Login Component", () => {
  test("should match snapshot.", () => {
    const { container } = render(<Login />);

    expect(container).toMatchSnapshot();
  });

  test("should render the title 'Access your account'.", () => {
    render(<Login />);

    const title = screen.getByText("Acesse sua conta");

    expect(title).toBeInTheDocument();
  });

  test("should render the login form.", () => {
    render(<Login />);

    const loginInput = screen.getByPlaceholderText(
      "Digite seu CPF ou CNPJ cadastrado",
    );
    const passwordInput = screen.getByPlaceholderText("Digite sua senha");
    const submitButton = screen.getByRole("button", { name: /entrar/i });

    expect(loginInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  test("should render the SeaPay logo.", () => {
    render(<Login />);

    const logo = screen.getByAltText("Logo seaPay");

    expect(logo).toBeInTheDocument();
  });

  test("should render the registration link.", () => {
    render(<Login />);

    const signupText = screen.getByText(/Ainda não é membro\?/i);
    const signupLink = screen.getByText("Abra sua conta!");

    expect(signupText).toBeInTheDocument();
    expect(signupLink).toBeInTheDocument();
    expect(signupLink).toHaveAttribute("href", "/signup");
  });

  test("should render the background image.", () => {
    render(<Login />);

    const backgroundImage = screen.getByAltText("Pessoa sorrindo com celular");

    expect(backgroundImage).toBeInTheDocument();
  });

  test("should render the open account link.", () => {
    render(<Login />);

    const openAccountLink = screen.getByText("Abra sua conta!");
    expect(openAccountLink).toBeInTheDocument();
    expect(openAccountLink).toHaveAttribute("href", "/signup");
  });

  test("should send the user to the dashboard on successful login.", () => {
    const { useRouter } = require("next/navigation");
    const router = useRouter();

    render(<Login />);

    const loginInput = screen.getByPlaceholderText(
      "Digite seu CPF ou CNPJ cadastrado",
    );
    const passwordInput = screen.getByPlaceholderText("Digite sua senha");
    const submitButton = screen.getByRole("button", { name: /entrar/i });

    fireEvent.change(loginInput, { target: { value: "12345678900" } });
    fireEvent.change(passwordInput, { target: { value: "password123" } });

    fireEvent.click(submitButton);

    waitFor(() => {
      expect(router.push).toHaveBeenCalledWith("/dashboard");
    });
  });

  test("should show error message on failed login.", async () => {
    const toast = require("react-hot-toast").default;

    render(<Login />);

    const loginInput = screen.getByPlaceholderText(
      "Digite seu CPF ou CNPJ cadastrado",
    );

    const passwordInput = screen.getByPlaceholderText("Digite sua senha");
    const submitButton = screen.getByRole("button", { name: /entrar/i });

    fireEvent.change(loginInput, { target: { value: "wronguser" } });
    fireEvent.change(passwordInput, { target: { value: "wrongpassword" } });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(toast.error).toHaveBeenCalledWith("Login ou senha inválidos");
    });
  });
});
