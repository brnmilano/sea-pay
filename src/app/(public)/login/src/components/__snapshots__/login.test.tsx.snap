// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Login Component should match snapshot. 1`] = `
<div>
  <main
    class="containerLogin"
  >
    <section
      class="loginSection"
    >
      <div
        class="loginContent"
      >
        <img
          alt="Logo seaPay"
          class="logo"
          height="42"
          src="[object Object]"
          width="114"
        />
        <h1
          class="title"
        >
          Acesse sua conta
        </h1>
        <form
          class="loginForm"
        >
          <div
            class="inputContainer"
          >
            <div
              class="container undefined"
            >
              <label
                class="label"
                for="«r0»"
              >
                Login
              </label>
              <div
                class="content"
              >
                <input
                  aria-description="Campo de login, digite seu CPF ou CNPJ cadastrado"
                  class="input undefined"
                  id="«r0»"
                  name="login"
                  placeholder="Digite seu CPF ou CNPJ cadastrado"
                  type="text"
                />
              </div>
              <div
                class="errorMessage undefined"
              >
                <p />
              </div>
            </div>
            <div
              class="container undefined"
            >
              <label
                class="label"
                for="«r1»"
              >
                Senha
              </label>
              <div
                class="content"
              >
                <input
                  aria-description="Campo de senha, digite sua senha"
                  class="input undefined"
                  id="«r1»"
                  name="password"
                  placeholder="Digite sua senha"
                  type="password"
                />
                <span
                  aria-controls="«r1»"
                  aria-label="Mostrar senha"
                  aria-pressed="false"
                  class="icon clickable"
                  role="button"
                  tabindex="0"
                  title="Mostrar senha"
                >
                  <svg
                    aria-hidden="true"
                    class="lucide lucide-lock"
                    fill="none"
                    height="20"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <rect
                      height="11"
                      rx="2"
                      ry="2"
                      width="18"
                      x="3"
                      y="11"
                    />
                    <path
                      d="M7 11V7a5 5 0 0 1 10 0v4"
                    />
                  </svg>
                </span>
              </div>
              <div
                class="errorMessage undefined"
              >
                <p />
              </div>
            </div>
          </div>
          <button
            aria-label="Entrar"
            class="button primary size-large"
            data-size="large"
            data-variant="primary"
          >
            Entrar
          </button>
        </form>
        <p
          class="signupText"
        >
          Ainda não é membro?
           
          <a
            class="signupLink"
            href="/signup"
          >
            Abra sua conta!
          </a>
        </p>
      </div>
    </section>
    <aside
      class="loginImage"
    >
      <img
        alt="Pessoa sorrindo com celular"
        src="[object Object]"
      />
    </aside>
  </main>
</div>
`;
